### PR TITLE
Change public objects logic & Search with ACL control

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 1]
```diff
@@ -172,6 +172,9 @@ export class WorkspaceSavedObjectsClientWrapper {
     const findWithWorkspacePermissionControl = async <T = unknown>(
       options: SavedObjectsFindOptions
     ) => {
+      if (options.type === WORKSPACE_TYPE) {
+        return await wrapperOptions.client.find<T>(options);
+      }
       if (options.workspaces) {
         options.workspaces = options.workspaces.filter(
```
Comment (ruanyl) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:185]: hmm, I think `array.filter()` with `async/await` callback won't work as expected here, we need to use `for...of`

Comment (SuZhou-Joe) [reply src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:185]: Agree, this logic need to be changed when ACL class in ready.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 2]
```diff
@@ -214,26 +214,57 @@ export class WorkspaceSavedObjectsClientWrapper {
...
-          options.workspaces = permittedWorkspaceIds;
-          options.queryDSL = queryDSL;
+          options.workspaces = undefined;
+          /**
+           * Select all the docs that
+           * 1. ACL matches right or write permission OR
+           * 2. workspaces matches library_read or library_write or management OR
+           * 3. Records without workspaces field (Advances settings)
+           */
+          options.queryDSL = {
```
Comment (SuZhou-Joe) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:235]: @gaobinlong Please help to review this query DSL, I do not know if it is easy to read and with good search performance.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 3]
```diff
@@ -214,26 +214,57 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+            query: {
+              bool: {
+                filter: [
+                  {
+                    bool: {
+                      should: [
+                        {
+                          bool: {
```
Comment (gaobinlong) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:243]: This condition: `type: 'config'` , we can use a single term query instead, `bool.must` is not needed, they have same effect.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 4]
```diff
@@ -214,26 +214,57 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+                              },
+                            },
+                          },
+                        },
+                        queryDSL.query,
+                        {
+                          bool: {
+                            should: permittedWorkspaceIds?.map((item) => ({
+                              terms: {
+                                workspaces: [item],
```
Comment (gaobinlong) [inline src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts:256]: This condition ` workspaces: [item]`, `bool.should` is not needed, use single terms query instead: 
```
terms:{
workspaces: permittedWorkspaceIds
}
```

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 3 above)
Comment (SuZhou-Joe) [reply src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts]: Thanks and done.

File: src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 4 above)
Comment (SuZhou-Joe) [reply src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts]: Thanks and done.

File: src/plugins/workspace/public/plugin.ts [hunk 5]
```diff
@@ -73,7 +78,7 @@ export class WorkspacesPlugin implements Plugin<{}, {}, WorkspacesPluginSetupDep
     /**
      * Retrieve workspace id from url
      */
-    const workspaceId = this.getWorkspaceIdFromURL();
+    const workspaceId = this.getWorkspaceIdFromURL() || PUBLIC_WORKSPACE;
```
Comment (ruanyl) [inline src/plugins/workspace/public/plugin.ts:81]: I'm wondering if we need to make `PUBLIC_WORKSPACE` a default workspace id here? For public workspace, it should have its own url like `/w/public` which can be retrieved from url as well.

File: src/plugins/workspace/public/plugin.ts [hunk 6]
```diff
@@ -194,7 +199,9 @@ export class WorkspacesPlugin implements Plugin<{}, {}, WorkspacesPluginSetupDep
     if (this.coreStart) {
       return this.coreStart.workspaces.client.currentWorkspaceId$.subscribe(
         (currentWorkspaceId) => {
-          this.coreStart?.savedObjects.client.setCurrentWorkspace(currentWorkspaceId);
+          this.coreStart?.savedObjects.client.setCurrentWorkspace(
+            currentWorkspaceId || PUBLIC_WORKSPACE
```
Comment (ruanyl) [inline src/plugins/workspace/public/plugin.ts:203]: I feel url should be the single source of truth, if need to enter to `public` workspace, it should navigate to /w/public`

Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.ts]: Yes, after discussion on some special types of workspaces, we should remove this default logic here.

File: src/core/public/saved_objects/saved_objects_client.ts [hunk 7]
```diff
@@ -366,14 +377,24 @@ export class SavedObjectsClient {
...
-      ...(options.workspaces || [await this._getCurrentWorkspace()]),
-      'public',
-    ].filter((item) => item);
+    const currentWorkspaceId = this._getCurrentWorkspace();
+    let finalWorkspaces;
+    if (options.hasOwnProperty('workspaces')) {
+      finalWorkspaces = options.workspaces;
+    } else if (typeof currentWorkspaceId === 'string') {
+      finalWorkspaces =
+        currentWorkspaceId === PUBLIC_WORKSPACE
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/public/saved_objects/saved_objects_client.ts]: Change the logic here, I thought `public` workspace will be the place to display all the objects an user have access to.

File: src/plugins/workspace/public/plugin.ts (hunk 5 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.ts]: Agree, change the logic here.

File: src/core/public/workspace/workspaces_client.ts [hunk 8]
```diff
@@ -70,6 +70,8 @@ export class WorkspacesClient {
         }
       }
     );
+
+    this.init();
```
Comment (ruanyl) [inline src/core/public/workspace/workspaces_client.ts:74]: I originally moved the calling of `init()` to workspace plugin `setup()` function so it only calls `init()` when workspace is enabled. Any reason to move it to `constructor()`?

**Review by ruanyl — APPROVED** [review]

**Pull request merged**